botocore/client.py

```diff
@@ -106,22 +108,25 @@ def _register_retries(self, service_model):
         self._event_emitter.register('needs-retry.%s' % endpoint_prefix,
                                      handler, unique_id=unique_id)
 
-    def _get_signature_version_and_region(self, service_model, region_name,
-                                          is_secure, scoped_config,
-                                          endpoint_url):
-        # Get endpoint heuristic overrides before creating the
-        # request signer.
-        resolver = self._endpoint_resolver
-        scheme = 'https' if is_secure else 'http'
-        endpoint_config = resolver.construct_endpoint(
-            service_model.endpoint_prefix, region_name, scheme=scheme)
-
-        # Signature version override from endpoint.
+    def _get_preferred_signature_version(self, versions):
+        # Use signers in the following preference order.
+        preference_order = ['s3', 's3v4', 'v4', 'v2']
```

Comment:
The endpoint resolver will provide all of the metadata about an endpoint, including available signature versions. For example, `aws-cn` regions only support `s3v4` while most `aws` regions support `s3` and `s3v4`. Higher level code that consumes this endpoint metadata needs to make a choice as to how to use that metadata. In Boto3 and the CLI, we currently always pick `s3` over `s3v4`.

I'm open to suggestions on where we pick which signature version to use.